src/engine.ts

```diff
@@ -360,7 +360,7 @@ export class Engine {
     return result.map((output) => {
       const status = TransactionStatus.decode(output);
       if (status.success !== undefined) return status.success.output;
-      else if (status.revert !== undefined) return Err(status.revert);
+      else if (status.revert !== undefined) return status.revert.output;
```

Comment:
I don't like that this means revert are indistinguishable from successes. I would think the user should care if the call succeeded or not.

Reply:
The dapps usually look for the `reason` or the revert messages to debug/troubleshoot and make sure that the smart contract working as expected.

What I am trying is to give the user the same experience like Geth. 

Before:
```
reason: undefined,
  hijackedStack: 'StatusError: Transaction: 0x06ef001a5514e0918d0146e95b4fc815bf845df982714ab059bfb42866f6e2fb exited with an error (status 0). \n' +
    '     Please check that the transaction:\n' +
    '     - satisfies all conditions set by Solidity `require` statements.\n' +
    '     - does not trigger a Solidity `revert` statement.\n' +
    '\n' +
    '    at Object.receipt (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/handlers.js:128:1)\n' +
    '    at runMicrotasks (<anonymous>)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:97:5)\n' +
    '    at Function.start (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/override.js:49:1)'
}
```

After:

```
  reason: 'Invalid Transfer',
  hijackedStack: 'StatusError: Transaction: 0x31779de9595405bf73c8842912ed1ad4f4553e621551c0c4d97c861ee658ee34 exited with an error (status 0). Reason given: Invalid Transfer.\n' +
    '     Please check that the transaction:\n' +
    '     - satisfies all conditions set by Solidity `require` statements.\n' +
    '     - does not trigger a Solidity `revert` statement.\n' +
    '\n' +
    '    at Object.receipt (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/handlers.js:128:1)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:97:5)\n' +
    '    at Function.start (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/override.js:49:1)'
}
```

Reply:
I must be missing something. I don't understand how the caller of this `view` function knows the transaction returned `revert` if in both the success and revert cases only raw bytes are returned?

Reply:
What I got from this view is a [borsh deserialized transaction status](https://github.com/aurora-is-near/aurora.js/blob/9845b0fd5ef179b589ace2ca981f2c0434765667/src/schema.ts#L121)

Reply:
I second @birchmd's concern. Also, success returns should be wrapped in `Ok(...)`, no?

Reply:
@birchmd I would think that the original code didn't give useful results since `Err(status.revert)` is wrapping an object, not a byte array or string. Instead of `status.revert.output`, could we do `Err(status.revert.output)`?

Reply:
If I do it as shown below:
```javascript
...
      if (status.success !== undefined) return Ok(status.success.output);
      else if (status.revert !== undefined) return Err(status.revert.output);
....
```
I got `undefined` reason, this mean that it doesn't throw the errors:
```
  reason: undefined,
  hijackedStack: 'StatusError: Transaction: 0x5869d76270c300a500110db1eff61febd3140b2c5a8836aaa1e1872ec5010bde exited with an error (status 0). \n' +
    '     Please check that the transaction:\n' +
    '     - satisfies all conditions set by Solidity `require` statements.\n' +
    '     - does not trigger a Solidity `revert` statement.\n' +
    '\n' +
    '    at Object.receipt (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/handlers.js:128:1)\n' +
    '    at runMicrotasks (<anonymous>)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:97:5)\n' +
    '    at Function.start (/usr/local/lib/node_modules/truffle/build/webpack:/packages/contract/lib/override.js:49:1)'
}```

Reply:
> This is what I got from this view is a [borsh deserialized transaction status](https://github.com/aurora-is-near/aurora.js/blob/9845b0fd5ef179b589ace2ca981f2c0434765667/src/schema.ts#L121)

Sorry, I should have been more clear. I do understand the return value from the Aurora contract's `view` call is a Borsh-serialized `TransactionStatus` object. What I was asking about was the `view` function here in this JS library. It calls the Aurora contract's function, then deserializes the `TransactionStatus`, all that looks good, but then in the revert case it acts the same as in the success case, which does not make sense to me.

I agree with @artob that a better way to write this would be to return a type `Promise<Result<Uint8Array, any>>` and obtain this using `andThen`

```javascript
    return result.andThen((output) => {
      const status = TransactionStatus.decode(output);
      if (status.success !== undefined) return Ok(status.success.output);
      else if (status.revert !== undefined) return Err(status.revert);
      else if (status.outOfGas !== undefined) return Err(status.outOfGas);
      else if (status.outOfFund !== undefined) return Err(status.outOfFund);
      else if (status.outOfOffset !== undefined) return Err(status.outOfOffset);
      else if (status.callTooDeep !== undefined) return Err(status.callTooDeep);
      else return Err('Failed to retrieve data from the contract');
    });
```

I do not understand why @artob thinks returning an object from this code is not a useful result though. Could you elaborate on that more? How is this function used in other libraries? For example [in the CLI we simply unwrap and show the bytes](https://github.com/aurora-is-near/aurora-cli/blob/master/src/aurora.ts#L190-L201) and I would expect the revert object would be handled fine in the unwrap.